### ISSUE TITLE
Can't pass environment variables to `configure_file()`

### ISSUE THREAD

**Issue by bgilbert** [mesonbuild/meson#12493]
**Describe the bug**
Unlike other functions that invoke programs, `configure_file()` doesn't have an `env` argument for passing environment variables to the invoked `command`.

**To Reproduce**
```
project('test')
env = environment()
env.set('FOO', 'bar')
configure_file(
  capture : true,
  command : 'printenv',
  env : env,
  output : 'outfile'
)
```
**Expected behavior**
`meson setup` succeeds and `outfile` contains `FOO=bar`.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
native
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)
Fedora 39
* what Python version are you using e.g. 3.8.0
3.12.0
* what `meson --version`
1.2.3
* what `ninja --version` if it's a Ninja build
1.11.1

**Comment by bgilbert** [comment]
Well, the `meson.add_*_script()` methods don't take environment variables either.

**Comment by eli-schwartz** [comment]
I'm not necessarily averse to adding them, but I'm curious to hear more about the real world need. The locations where it has been added thus far was in response to such need -- in particular, it's often necessary to convey information to third-party programs that assume the use of environment variables.

But it seems uncommon for any of this to matter for install scripts, which are usually shell or python and custom written for the project + shipped in the project source tree. In such cases it's not really a problem to write the script from the start to accept command line arguments, excepting the use of DESTDIR which meson itself takes charge of.

configure_file commands are similar, but with the added nuance of it usually being a bad idea to use commands there anyway as configure time commands are usually (but not always!) just as easily done at build time with the added benefit of not forcing a reconfigure every time the input file changes.

tl;dr this was never added before, because everyone sort of assumed it had no practical use and therefore didn't bother.

**Comment by bgilbert** [comment]
I don't have an immediate use for the `meson.add_*_script` case, and it turns out `configure_file()` isn't going to work anyway.  My script needs to run `meson introspect`, which understandably isn't available when `configure_file()` runs.

I still think the functionality would be useful though.  In the same vein as #12492, it's difficult to write a consistent set of build helpers when different invocation paths have different limitations.  In my case, `meson.build` populates a set of environment variables from `find_program()` calls so the helpers know how to run tools like `objdump`.  It'd be nice to pass that single environment block to every helper invocation without navigating multiple calling conventions.

**Comment by fvalette** [comment]
> I'm not necessarily averse to adding them, but I'm curious to hear more about the real world need. The locations where it has been added thus far was in response to such need -- in particular, it's often necessary to convey information to third-party programs that assume the use of environment variables.
> 
> But it seems uncommon for any of this to matter for install scripts, which are usually shell or python and custom written for the project + shipped in the project source tree. In such cases it's not really a problem to write the script from the start to accept command line arguments, excepting the use of DESTDIR which meson itself takes charge of.
> 
> configure_file commands are similar, but with the added nuance of it usually being a bad idea to use commands there anyway as configure time commands are usually (but not always!) just as easily done at build time with the added benefit of not forcing a reconfigure every time the input file changes.
> 
> tl;dr this was never added before, because everyone sort of assumed it had no practical use and therefore didn't bother.

Hi @eli-schwartz,
I've got a "real world" use case for env var in configure_file. I use Kconfig + Meson for one project, and i'd like to run `oldconfig` at meson setup time (or `defconfig` if one uses a defconfig file instead of a full config file). `Kconfig` (and and python `Kconfiglib`) heavily use environment variables (e.g. `KCONFIG_CONFIG` for outputted config file, `srctree`, and so on). This step cannot be a custom_target because the config file is used as `keyval` as an input for `confguration_data` dict, nor a `run_command` because a depfile is produced in order to reconfigure the project if any of the parsed Kconfig files changed, or the `.config` (e.g. after running `menuconfig`).

I worked around this issue using an inline python script the export those variables before calling `oldconfig` and/or `defconfig` at setup time (but i use devenv feature to export those variable to the developer once the project is configured for `menuconfig` and `savedefconfig` for instance).
